src/categorical_algebra/Sets.jl

```diff
@@ -191,12 +297,71 @@ end
   end
 end
 
-@inline compose_id(f::SetFunction, g::SetFunction) = do_compose(f, g)
+@inline compose_id(f::SetFunction, g::SetFunction) = compose_inv(f, g)
 @inline compose_id(f::SetFunction, ::IdentityFunction) = f
 @inline compose_id(::IdentityFunction, g::SetFunction) = g
 @inline compose_id(f::IdentityFunction, ::IdentityFunction) = f
 
-do_compose(f::SetFunction, g::SetFunction) = CompositeFunction(f, g)
+# Developer's note: `compose_inv` and its helper `comp_inv_helper` are
+# introduced here to ensure that `f` and `inv(f)` will compose to the
+# appropriate `IdentityFunction` object when they meet. In this, the type
+# `BijectionBimap` plays the decisive role. It provides an effective
+# representation of a function whose inverse is known. But note, the same
+# behavior is not guaranteed for some instances of `FinBijection`s. That is,
+# composition of inverses is not guaranteed to result in an `IdentityFunction`.
+# See test/categorical_algebra/FinSets.jl.
+#
+# The implementation of `compose_inv` is somewhat tortured, and it is due in
+# part to a disadvantage of the extant `CompositeFunction` type. Iterated
+# composites are represented essentially as a tree of `SetFunction`s. As a
+# consequence, it is not possible to tell from the type of a `CompositeFunction`
+# whether the leftmost (or rightmost) item of the composite is a
+# `BijectionBimap`. Because of all this, the implementation introduces an
+# efficiency loss whenever using `compose` with a `CompositeFunction`, which
+# grows worse as the "tree" of composites grows deeper. To make things simpler
+# for `compose_inv`, the obvious idea would be to try refactoring
+# `CompositeFunction` to store a list of `SetFunction`s instead of a pair.
+
+function compose_inv(f::SetFunction, g::SetFunction)
```

Comment:
I think this might be easier with MLStyle pattern matching and structural recursion.